zxtune: init at r5054

EBADBEEF opened the pull request:
## Description of changes

zxtune is a cross-platform chiptunes music player.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

EBADBEEF committed:
maintainers: add EBADBEEF

EBADBEEF committed:
zxtune: init at r5055

FliegendeWurst commented on pkgs/applications/audio/zxtune/default.nix:
```diff
@@ -0,0 +1,129 @@
+in with lib;
+stdenv.mkDerivation rec {
+  pname = "zxtune";
+  version = "r5054";
```
```suggestion
  version = "5054";
```
Usually the version attribute starts with a digit.

EBADBEEF replied:
Done.

FliegendeWurst commented on pkgs/applications/audio/zxtune/default.nix:
```diff
@@ -0,0 +1,129 @@
+  src = fetchFromBitbucket {
+    owner = "zxtune";
+    repo = "zxtune";
+    rev = "${version}";
```
```suggestion
    rev = "r${version}";
```

EBADBEEF replied:
Done.

FliegendeWurst commented on pkgs/applications/audio/zxtune/default.nix:
```diff
@@ -0,0 +1,129 @@
+  platformName = "linux";
+  staticBuildInputs = [ boost zlib ]
+    ++ lib.optional withQt (if (supportWayland) then qt5.qtwayland else qt5.qtbase);
+in with lib;
```
I'm not sure, but I believe `with lib;` should be avoided in front of package definitions. This one uses only a few `lib.*` functions.

FliegendeWurst reviewed (commented):
Thanks for packaging zxtune!

Package builds and zxtune-qt launches.

Result of `nixpkgs-review pr 265519` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>zxtune</li>
  </ul>
</details>

EBADBEEF commented:
Thanks for the review!

FliegendeWurst commented on pkgs/applications/audio/zxtune/default.nix:
```diff
@@ -0,0 +1,128 @@
+    '';
+    homepage = "https://zxtune.bitbucket.io/";
+    license = licenses.gpl3;
+    platforms = platforms.linux;
```
This is set to Linux because you can't test on Darwin, right?

EBADBEEF replied:
Exactly. I suspect there will need to be some more tricks in the makefile stuff.

FliegendeWurst replied:
Good to know (and perfectly acceptable, most maintainers can't test on Darwin). It was a bit confusing to see all those `isLinux` / `isDarwin` checks and then this line at the end.

I'd put a short comment above the "platforms =" line so future readers are not confused.

FliegendeWurst commented on pkgs/applications/audio/zxtune/default.nix:
```diff
@@ -0,0 +1,128 @@
+      "support_${name}=" + (if (var) then "1" else "");
+    makeOptsCommon = [
+      ''-j$NIX_BUILD_CORES''
+      ''root.version=${version}''
```
`r${version}`?

EBADBEEF replied:
Yeah, should be, will do.

FliegendeWurst reviewed (approved):
Result of `nixpkgs-review pr 265519` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>zxtune</li>
  </ul>
</details>

I'm assuming you tested the off-by-default file/audio backends.

EBADBEEF commented:
> Result of `nixpkgs-review pr 265519` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
> 
> 1 package built:
> I'm assuming you tested the off-by-default file/audio backends.

Uh oh. You caught me! It doesn't build with SDL2 enabled.

EBADBEEF commented:
OK, I think I have addressed all the review feedback now! Thanks for the thoughtful comments.
- tested build and functionality with all optional flags
- added comment about darwin/windows platform support

FliegendeWurst reviewed (approved):
Result of `nixpkgs-review pr 265519` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>zxtune</li>
  </ul>
</details>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2919

pbsds commented on pkgs/applications/audio/zxtune/default.nix:
```diff
@@ -0,0 +1,136 @@
+  preFixup = lib.optionalString withQt ''
+    wrapQtApp "$out/bin/zxtune-qt"
+  '';
+
```
```suggestion

  passthru.updateScript = nix-update-script {
    attrPath = "zxtune";
    extraArgs = [ "--version-regex" "r(.*)" ];
  };

```

pbsds commented on pkgs/applications/audio/zxtune/default.nix:
```diff
@@ -0,0 +1,136 @@
+{ lib
+, stdenv
+, fetchFromBitbucket
+, boost
```
```suggestion
, nix-update-script
, boost
```

pbsds commented on pkgs/applications/audio/zxtune/default.nix:
```diff
@@ -0,0 +1,136 @@
+  ];
+
+  postPatch = ''
+    sed -e 's@OpenAL/@AL/@g' -i src/sound/backends/gates/openal_api.h
```
```suggestion
    # a comment why this is needed
    substituteInPlace src/sound/backends/gates/openal_api.h \
      --replace "OpenAL/" "AL/"
```

pbsds commented on pkgs/applications/audio/zxtune/default.nix:
```diff
@@ -0,0 +1,136 @@
+      ''tools.rcc=${qt5.qtbase.dev}/bin/rcc''
+    ];
+  in ''
+    make ${builtins.toString makeOptsCommon} -C apps/xtractor
+    make ${builtins.toString makeOptsCommon} -C apps/zxtune123
+  '' + lib.optionalString withQt ''
+    make ${builtins.toString (makeOptsCommon ++ makeOptsQt)} -C apps/zxtune-qt
```
```suggestion
    runHook preBuild
    make ${builtins.toString makeOptsCommon} -C apps/xtractor
    make ${builtins.toString makeOptsCommon} -C apps/zxtune123
  '' + lib.optionalString withQt ''
    make ${builtins.toString (makeOptsCommon ++ makeOptsQt)} -C apps/zxtune-qt
  '' + ''
    runHook postBuild
```

pbsds reviewed (commented):
Result of `nixpkgs-review pr 265519` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>zxtune</li>
  </ul>
</details>

Welcome to nixpkgs! This one is looking pretty good!

Runs fine on my end, just some small nits and improvements

EBADBEEF commented:
Changes: 
- Rebased (darn it is hard to track my changes with this, sorry... not really sure the best way to work with github here)
- Moved to pkgs/by-name
- Enabled flac, mp3, vorbis by default (it's already a big package, might as well add this functionality)
- Followed @pbsds review suggestions (build hooks, substituteInPlace instead of sed, nix-update)
- Regarding nix-update... I didn't know that existed, thanks! However, it doesn't support bitbucket! Hopefully that will change soon => https://github.com/Mic92/nix-update/pull/208 ;-) Also noticed I can't use "refs/tags" syntax for rev with nix-update.

pbsds commented:
@ofborg eval

pbsds commented on pkgs/by-name/zx/zxtune/package.nix:
```diff
@@ -0,0 +1,146 @@
+    longDescription = ''
+      Player of computer music from ZX Spectrum, Amstrad, Sam Coupe, PC, Amiga,
+      Atari, Acorn, C64, SNES, Nes, Sega Master System, GameBoy, TurboGrafX,
+      MSX, NDS, and more.
```
```suggestion
      MSX, NDS, and more. Powered by vgmstream.
```

For the sake of discoverability

pbsds replied:
_(ofborg seems to need the fore push anyway)_

EBADBEEF replied:
Nice idea. Updated description and included other notable libraries. Also bumped version to r5055.
